src/utils.js

```diff
@@ -0,0 +1,110 @@
+}
+
+export const rando = () =>{
+        const colors = ['purple', 'orange', 'red', 'green', 'yellow', 'blue', 'white', 'blue']
+        let randomColor = colors[Math.floor(Math.random()*colors.length)];
+        return randomColor
+     }
+
+
+export const find_next_non_null = (previousState,rowIndex, column) =>{
+        for(let i=rowIndex-1; i >= 0; i--){
+            if(previousState[i][column] !==null){
+                return i
+            }
+        }
+    }
+
+export const findMatches = (inputSquares) => {
+    //vertical check
+    let squares = [...inputSquares]
+    let toChange = []
+    for(let row = squares.length-1; row >= 0 ; row--) {
+        for(let column = 0; column < squares[row].length; column++) {
+            console.log(squares[row][column])
+            if(squares[row + 1] && squares[row + 2]){
+                if(squares[row + 1][column] && squares[row + 2][column]){
+                    if (squares[row][column]=== squares[row + 1][column] && squares[row][column]===squares[row + 2][column]){
+                        toChange.push({row: row, column: column})
+                        toChange.push({row: row + 1, column:column})
+                        toChange.push({row: row + 2, column: column})
+                    }
+                }
+            }
+
+            if(squares[row - 1] && squares[row - 2]){
+                if(squares[row - 1][column] && squares[row - 2][column]){
+                    if (squares[row][column]=== squares[row - 1][column] && squares[row][column]===squares[row - 2][column]){
+                        toChange.push({row: row, column: column})
+                        toChange.push({row: row - 1, column:column})
+                        toChange.push({row: row - 2, column: column})
+                    }
+                }
+            }
+
+                //horizontal check
+                if(squares[row][column +1] && squares[row ][column +2]){
+                    if(squares[row][column] === squares[row][column +1] && squares[row][column]===squares[row ][column +2]){
+                        toChange.push({row: row, column: column })
+                        toChange.push({row: row, column:column + 1})
+                        toChange.push({row: row, column: column + 2})
+                    }
+                }
+
+                if(squares[row][column -1] && squares[row ][column -2]){
+                    if(squares[row][column] === squares[row][column -1] && squares[row][column]===squares[row ][column -2]){
+                        toChange.push({row: row, column: column })
+                        toChange.push({row: row, column:column - 1})
+                        toChange.push({row: row, column: column - 2})
+                    }
+                }
+        }
+    }
+
+    if(toChange.length ===0){
+        return null
+    }
+    
+    squares = null_matches(squares, toChange)
+
+        return squares       
+}
+
+
+
+export const null_matches =(squares, toChange)=>{
```

Comment:
Structuring this code so that we set cells to be removed to null totally works, but has a few shortcomings. I would consider an approach where you have a function that returns a list of all groups of indices that are a chain of 3 or more, and a separate function that takes those, an existing board state, and creates a new board state with new colors added appropriately. 

Nulling the cells out means that, well, they might be null at different times. If cells are never null, the null checks on lines 41, 51, 62 and 70 are unnecessary, which means that the pretty complicated block of logic to check for matches can get somewhat simpler.